### PR TITLE
Enable IAM User Deletion

### PR DESCRIPTION
**Pull request by zined**
EDIT: opened this early to get some feedback, very much appreciated!

Coming from a setup with a SAML Identity Provider, which means no human users inside of IAM, I want to keep the list of Machine-Users inside IAM as documented and explicit as possible, and enforce the state.

This Pull-Request adds support to delete IAM Users and a filter on IAM Users that does not rely on the CredentialReport with it's 4h delay.

This is still missing tests, I'll have to get familiar with the nosetest/placebo setup first. If someone would want to remote pair this, I'd be happy to!

### CODE REVIEW

**Comment by zined** [comment]
rebased against master

File: c7n/resources/iam.py [hunk 1]
```diff
@@ -774,6 +774,120 @@ def _user_mfa_devices(resource):
         return matched
 
 
+@User.filter_registry.register('username')
+class UserName(ValueFilter):
```
Comment (kapilt) [inline c7n/resources/iam.py]: unclear this is needed, the default value filter on username can address subattributes directly, the key in a value filters is a jmepath expression that allows for this directly.

File: c7n/resources/iam.py [hunk 2]
```diff
@@ -774,6 +774,120 @@ def _user_mfa_devices(resource):
...
+                        AccessKeyId=access_key['AccessKeyId'])
+
+            self.log.debug('Detaching UserPolicies from user "%s"' % (r['UserName']))
+            response = client.list_attached_user_policies(UserName=r['UserName'])
+            for user_policy in response['AttachedPolicies']:
+                client.detach_user_policy(UserName=r['UserName'],
+                        PolicyArn=user_policy['PolicyArn'])
+
+            self.log.debug('Deleting user "%s"' % (r['UserName']))
+            client.delete_user(UserName=r['UserName'])
```
Comment (kapilt) [inline c7n/resources/iam.py]: if you just delete the user, do you really need to do any of the other activities here, ie a deleted user doesn't have attached user policies or valid access keys i would assume.

File: c7n/resources/iam.py (hunk 1 above)
Comment (zined) [inline c7n/resources/iam.py]: @kapilt thanks for the feedback! I'll try to see if I can filter the same way using a JMESPath in the key field.

File: c7n/resources/iam.py (hunk 2 above)
Comment (zined) [inline c7n/resources/iam.py]: Yeah, unfortunately you have to take care of deleting/detaching those entities before you can reliably delete a user. I've even (for sure) missed MFA Tokens in here, and probably also Group Memberships.. :)

quoting from http://docs.aws.amazon.com/IAM/latest/APIReference/API_DeleteUser.html

```
DeleteUser

Deletes the specified IAM user. The user must not belong to any groups or have any access keys, signing certificates, or attached policies.
```

**Comment by zined** [comment]
@kapilt Thanks a lot for your review and your willingness to remote pair on this. I'm located in CEST (UTC+2, EST+7). Would something between 0900 EST and 1200 EST work for you? Also, any preferrences on how to remote pair? I've been using Screenhero some times in the past, or old-school SSH/Screen/TMUX based Setups, but am positive to try anything else that works for you.

File: c7n/resources/iam.py (hunk 1 above)
Comment (zined) [inline c7n/resources/iam.py]: you're actually completely right. the following snippet acchieves the same, without adding a new filter:

```
policies:
  - name: defined-machine-users
    resource: iam-user
    filters:
      - type: value
        key: "Arn"
        op: not-in
        value:
          - arn:aws:iam::123456789012:user/valid-user-1
          - arn:aws:iam::123456789012:user/valid-user-2
```

File: c7n/resources/iam.py (hunk 2 above)
Comment (zined) [inline c7n/resources/iam.py]: double checked. both group memberships and mfa devices need to be taken care of before the user can be deleted. i hope i've caught everything now.

**Comment by kapilt** [comment]
i'm game for 11-12 EST if that still works (short notice), tmux/ssh works for me.

**Comment by zined** [comment]
> i'm game for 11-12 EST if that still works (short notice), tmux/ssh works for me.

followed you on twitter for pairing "key-exchange" ;-)

**Comment by kapilt** [comment]
fwiw, github does nice as repo of public keys ala `curl https://github.com/zined.keys >> ~/.ssh/authorized_keys`

**Comment by kapilt** [comment]
i can schedule again for wednesday, might be good to address the extant review comments if you have time before that.

**Comment by kapilt** [comment]
message me on gitter re timing wednesday.

**Comment by kapilt** [comment]
I'm going to merge as is and cleanup post merge.

**Comment by kapilt** [comment]
one thing needed before we can merge this is getting a CLA signed per the project readme, direct link
https://docs.google.com/forms/d/19LpBBjykHPox18vrZvBbZUcK6gQTj7qv1O5hCduAZFU/viewform

**Comment by zined** [comment]
CLA signed.

**Comment by kapilt** [comment]
refactored to just delete the user, the dependent references are all yanked afaics automatically.

**Comment by kapilt** [comment]
github status checks having some issues, upstream services all report good though.

**Comment by kapilt** [comment]
so i totally missed the dependency cleanup here it looks like, and it is indeed needed.

**Comment by zined** [comment]
> so i totally missed the dependency cleanup here it looks like, and it is indeed needed.

you mean https://github.com/capitalone/cloud-custodian/pull/1300/commits/658bd0b9ec999b38d58ad716e6e13e721b2f6eb0 ?

**Pull request merged**